File: packages/dhis2w-fhir-serve/frontend/src/lib/health.ts
/**
 * What `GET /facade/metadata-health` answers: what the DHIS2 instance behind this run holds that the guide
 * cannot carry cleanly, and how far the selection is translated.
 *
 * Not FHIR. There is no FHIR shape for "this DHIS2 name has a `<` in it" and none at all for a
 * translation that stops halfway, so the server answers plain JSON and these interfaces are its
 * models, field for field, in the wire's own spelling. The Python side is
 * `dhis2w_fhir_serve.health`.
 *
 * THE WORDING IS THE SERVER'S. Every finding carries the sentence `d2w fhir validate` wrote for it,
 * and nothing here rephrases one: the same defect read in a terminal and read in a browser has to be
 * the same sentence, or two people looking at one instance are looking at two reports. What this
 * module adds is arrangement - which severity a row sits under, which object kind it is shelved
 * beside, and the arithmetic under the coverage strip.
 */

/** How bad one finding is, in the terms `d2w fhir validate` grades by. */
export type FindingSeverity = 'error' | 'warning' | 'info'

/** Whether a finding sits on an object this project publishes, or on one it never reads. */
export type FindingScope = 'selection' | 'instance'

/** The three severities in the order a page reads them: what stops a build first. */
export const FINDING_SEVERITIES: readonly FindingSeverity[] = ['error', 'warning', 'info'] as const

/** What each severity is called on screen - the word a reader scans for, capitalised as a label. */
export const SEVERITY_LABELS: Record<FindingSeverity, string> = {
    error: 'Errors',
    warning: 'Warnings',
    info: 'Notes',
}

/** What one of each is called in a sentence about a single row. */
export const SEVERITY_SINGULAR: Record<FindingSeverity, string> = {
    error: 'Error',
    warning: 'Warning',
    info: 'Note',
}

/**
 * The hue each severity wears, off the state tokens the receipts already use.
 *
 * A KIND GETS A HUE, AND THESE ARE THE SAME THREE KINDS THE REST OF THE APP PAINTS. A defect that
 * stops a build is the red a DHIS2 rejection wears, one that degrades a page is the amber a refused
 * submission wears, and a note about an object no build reads is not a state at all - so it gets the
 * muted pair rather than a third alarm colour. Words take the measured ink variant where a fill is
 * tuned for dots; see the note beside the tokens in index.css.
 */
export const SEVERITY_TINTS: Record<FindingSeverity, { dot: string; badge: string }> = {
    error: {
        dot: 'bg-status-rejected',
        badge: 'border-status-rejected/40 text-status-rejected bg-status-rejected/10',
    },
    warning: {
        dot: 'bg-status-refused',
        badge: 'border-status-refused/40 text-status-refused-ink bg-status-refused/10',
    },
    info: {
        dot: 'bg-muted-foreground',
        badge: 'border-border text-muted-foreground bg-muted/40',
    },
}

/** One thing the validator found about one DHIS2 object. */
export interface MetadataFinding {
    severity: FindingSeverity
    scope: FindingScope
    /** The validator's own name for the kind of defect - `invalid-code`, `template-hostile-name`. */
    category: string
    /** The DHIS2 metadata collection the object belongs to, in DHIS2's own spelling. */
    resource_type: string
    uid: string
    name: string
    code: string | null
    /** The DHIS2 field at fault, or null where the category is about none of the object's own fields. */
    field: string | null
    /** The exact problem, in the validator's own words. */
    message: string
    /** What this grade costs the project, in a sentence rather than in a severity word. */
    cost: string
}

/** How many findings there are of each severity. */
export interface FindingCounts {
    errors: number
    warnings: number
    infos: number
}

/** Whether a locale covers less than half the selection's translatable strings, or half or more. */
export type LocaleStanding = 'sparse' | 'majority'

/** One selected object that carries a translation in a locale little of the selection carries. */
export interface LocaleCarrier {
    resource_type: string
    uid: string
    name: string
    /** Whether this object carries a name translation in this locale. */
    carries_name: boolean
    /** Whether this object has a DHIS2 form name and carries a form-name translation in this locale. */
    carries_form_name: boolean
}

/** One selected object holding no translation in a locale most of the selection is translated into. */
export interface LocaleUntranslated {
    resource_type: string
    uid: string
    name: string
    name_untranslated: boolean
    form_name_untranslated: boolean
}

/**
 * How much of the selection one locale covers, and which side of it the server listed.
 *
 * `standing` says which of the two lists carries anything: a sparse locale is told through the
 * objects that carry it, a majority locale through the objects that do not. Neither list means
 * something different depending on the other, which is why there are two of them.
 */
export interface LocaleCoverage {
    /** The BCP-47 tag, normalised from the Java locale DHIS2 stores - `pt_BR` arrives as `pt-BR`. */
    locale: string
    /** Selected objects carrying a name translation in this locale. */
    name_count: number
    /** Selected objects that have a DHIS2 form name and carry a form-name translation in this locale. */
    form_name_count: number
    standing: LocaleStanding
    carriers: LocaleCarrier[]
    missing: LocaleUntranslated[]
}

/**
 * How far the selection is translated.
 *
 * `locales` is the union of the tags the selection's own translations carry, which is what "in use
 * on this instance" means: an instance is being maintained in the languages somebody wrote into it.
 * Empty is an instance nobody has translated, which is one language and a whole state.
 */
export interface TranslationCoverage {
    locales: string[]
    object_count: number
    /** Of those objects, how many DHIS2 gives a form name - the denominator the form-name counts read against. */
    form_named_count: number
    per_locale: LocaleCoverage[]
}

/**
 * The whole answer.
 *
 * `available` false is a compiled run and nothing else, and it arrives as a body rather than as a
 * refusal: there is nothing wrong with serving a compiled guide, and `reason` is the sentence the
 * page renders in place of the report.
 */
export interface MetadataHealth {
    available: boolean
    reason: string | null
    /** The `[generate] hostile_names` posture the severities were graded under, in the server's own line. */
    graded_under: string | null
    /** Metadata objects the validator swept, across every collection the instance holds. */
    object_count: number
    counts: FindingCounts
    findings: MetadataFinding[]
    translations: TranslationCoverage
}

/** What this page assumes before the read lands: an answer with nothing in it and no reason stated. */
export const EMPTY_METADATA_HEALTH: MetadataHealth = {
    available: true,
    reason: null,
    graded_under: null,
    object_count: 0,
    counts: { errors: 0, warnings: 0, infos: 0 },
    findings: [],
    translations: { locales: [], object_count: 0, form_named_count: 0, per_locale: [] },
}

/** How many findings one severity holds, off the counts the server stated. */
export function countOf(counts: FindingCounts, severity: FindingSeverity): number {
    if (severity === 'error') return counts.errors
    if (severity === 'warning') return counts.warnings
    return counts.infos
}

/**
 * Whether there is nothing on this run to put on a page.
 *
 * A finding of any severity is something to show. So is a locale in use, which is a fact about the
 * instance rather than a defect in it - the coverage a reader came to see is worth a page even on an
 * instance the validator had nothing to say about. Neither is an instance nobody has translated
 * whose names and codes all pass, and that is the one state this answers true for.
 */
export function isClean(health: MetadataHealth): boolean {
    return health.findings.length === 0 && health.translations.locales.length === 0
}

/** The one word the validator's `<field> <name> ...` shape continues with, which is what makes the cut safe. */
const NAMED_SUBJECT_VERB = 'contains '

/**
 * One finding's message as a row reads it, with the head the row's own columns already carry cut.
 *
 * THE VALIDATOR WRITES FOR A TERMINAL, where a finding is one line and has to name its subject:
 * `name CMC Post abortion related services - E&C/D&C contains '&' which the IG publisher ...`. A row
 * has an Object column and a Field column, so that head is the row's first two cells said a third
 * time, and it pushes the sentence a reader came for off the side of the table.
 *
 * THE CUT IS THE WHOLE SHAPE OR IT DOES NOT HAPPEN. It fires on the field label, the object's name,
 * and the verb that shape always continues with - all three, exactly - and the verb is what keeps it
 * honest: a data element named `Weight` whose form name is `Weight in kg` opens its finding with
 * `form name Weight in kg contains ...`, and matching on the name alone would cut it to `in kg
 * contains ...`. Anything the shape does not fit is rendered exactly as the validator wrote it, so a
 * change of wording upstream costs a head on a row rather than a mangled sentence.
 *
 * The wire payload is untouched either way: this is what the cell shows, not what the run found.
 */
export function findingMessage(finding: MetadataFinding): string {
    if (finding.field === null) return finding.message
    const head = `${finding.field} ${finding.name} ${NAMED_SUBJECT_VERB}`
    if (!finding.message.startsWith(head)) return finding.message
    return `Contains ${finding.message.slice(head.length)}`
}

/**
 * The findings one filter leaves, matched against what a reader would type.
 *
 * The object's name and its uid, because those are the two strings a person has in front of them -
 * a name they read in DHIS2, or a uid they copied out of a build log. Case-insensitive and a
 * substring rather than a prefix: half a name is what somebody types when they are not sure how the
 * instance spelled the rest of it.
 */
export function matchingFindings(findings: MetadataFinding[], query: string): MetadataFinding[] {
    return matchingObjects(findings, query)
}

/** One object kind's findings, under the kind DHIS2 calls it. */
export interface FindingGroup {
    /** The DHIS2 metadata collection, in DHIS2's own spelling - `dataElements`, `organisationUnits`. */
    resourceType: string
    findings: MetadataFinding[]
}

/** One severity's findings, shelved by the object kind each is about. */
export interface SeverityShelf {
    severity: FindingSeverity
    /** How many findings the shelf holds, over every group on it. */
    total: number
    groups: FindingGroup[]
}

/**
 * The findings arranged as the page reads them: severity first, then object kind.
 *
 * SEVERITY LEADS BECAUSE IT IS WHAT A READER CAME FOR. An error stops the build and a note does not,
 * so the errors are the first thing on the page whichever kind of object they are about; the kind is
 * the second cut, because a person fixing names fixes a run of data elements at once.
 *
 * A severity holding nothing is dropped rather than rendered empty, and so is a kind: a heading over
 * no rows is a heading that makes a reader look for rows. Groups are ordered by the DHIS2 collection
 * name, and the findings inside one keep the order the server sent - which is the validator's own
 * order, by severity then type then name.
 */
export function shelveFindings(findings: MetadataFinding[]): SeverityShelf[] {
    return FINDING_SEVERITIES.map((severity) => {
        const held = findings.filter((finding) => finding.severity === severity)
        return { severity, total: held.length, groups: groupByResourceType(held) }
    }).filter((shelf) => shelf.total > 0)
}

/** One severity's findings bucketed by object kind, kinds in DHIS2's own alphabetical order. */
export function groupByResourceType(findings: MetadataFinding[]): FindingGroup[] {
    const buckets = new Map<string, MetadataFinding[]>()
    for (const finding of findings) {
        const held = buckets.get(finding.resource_type)
        if (held === undefined) buckets.set(finding.resource_type, [finding])
        else held.push(finding)
    }
    return [...buckets.entries()]
        .toSorted((left, right) => left[0].localeCompare(right[0]))
        .map(([resourceType, held]) => ({ resourceType, findings: held }))
}

/**
 * One locale's coverage as the strip states it: how many of the selection's translatable strings it has.
 *
 * THE DENOMINATOR IS EVERY STRING, NOT EVERY OBJECT. A data element carries two strings DHIS2 puts on
 * a page - the name every vocabulary displays and the form name a question is asked under - and a
 * locale that translated the first and not the second has done half the work on that object. So the
 * total is the objects plus the form-named ones among them, and the covered count is the two counts
 * the server stated added together.
 *
 * A selection with nothing in it is 0 of 0, which reads as complete and is: there is nothing to
 * translate. Callers render the ratio rather than dividing by it.
 */
export interface CoverageRatio {
    locale: string
    covered: number
    total: number
    /** The share covered, 0 to 1, and 1 for a selection with nothing to translate. */
    share: number
    standing: LocaleStanding
    /** The objects that carry this locale, filled where it is sparse and empty where it is not. */
    carriers: LocaleCarrier[]
    /** The objects that do not, filled where it is a majority locale and empty where it is sparse. */
    missing: LocaleUntranslated[]
}

/** One locale's covered-over-total, over every translatable string in the selection. */
export function coverageRatio(coverage: TranslationCoverage, locale: LocaleCoverage): CoverageRatio {
    const total = coverage.object_count + coverage.form_named_count
    const covered = locale.name_count + locale.form_name_count
    return {
        locale: locale.locale,
        covered,
        total,
        share: total === 0 ? 1 : covered / total,
        standing: locale.standing,
        carriers: locale.carriers,
        missing: locale.missing,
    }
}

/** How many objects one locale's row lists - its carriers where it is sparse, the rest where it is not. */
export function listedCount(ratio: CoverageRatio): number {
    return ratio.standing === 'sparse' ? ratio.carriers.length : ratio.missing.length
}

/** Every locale's ratio, weakest first - the language somebody stopped translating leads the strip. */
export function coverageRatios(coverage: TranslationCoverage): CoverageRatio[] {
    return coverage.per_locale
        .map((locale) => coverageRatio(coverage, locale))
        .toSorted((left, right) => left.share - right.share || left.locale.localeCompare(right.locale))
}

/** One share as a whole percent, which is the only precision a strip of this size can carry. */
export function coveragePercent(ratio: CoverageRatio): number {
    return Math.round(ratio.share * 100)
}

/** What either translation list is a list of: an object a reader can find by name or by uid. */
export interface NamedObject {
    uid: string
    name: string
}

/**
 * The objects one filter leaves, matched the way the findings are - by object name and by uid.
 *
 * One filter box over the page rather than one per section: a reader looking for an object is
 * looking for it wherever it is, and a name that matched the findings and not the translation lists
 * would hide half of what is known about it.
 */
export function matchingObjects<T extends NamedObject>(objects: T[], query: string): T[] {
    const wanted = query.trim().toLowerCase()
    if (wanted === '') return objects
    return objects.filter(
        (object) => object.name.toLowerCase().includes(wanted) || object.uid.toLowerCase().includes(wanted),
    )
}

/**
 * Every locale's ratio with the objects it lists narrowed to the filter, the counts left as they are.
 *
 * The counts are what the instance holds and a filter does not change them - a meter that moved when
 * somebody typed would be answering a different question from the one it is labelled with.
 */
export function matchingRatios(coverage: TranslationCoverage, query: string): CoverageRatio[] {
    return coverageRatios(coverage).map((ratio) => ({
        ...ratio,
        carriers: matchingObjects(ratio.carriers, query),
        missing: matchingObjects(ratio.missing, query),
    }))
}

File: packages/dhis2w-fhir-serve/src/dhis2w_fhir_serve/health.py
"""What is not right about the DHIS2 metadata behind this run: the validate findings, plus translation coverage.

TWO ANALYSES, ONE ANSWER. The first is `d2w fhir validate` run over the connection this process
already holds - the same passes, the same graders, the same severity grading, and the same wording,
because a finding a reader acts on must not be phrased one way in a terminal and another way in a
browser. `dhis2w_fhir.service.validate_instance_codes` is the whole of it, and nothing here
re-implements a predicate that lives there.

The second is new here: how much of the selection is translated. DHIS2 holds a translation per
object, per property, per locale, and nothing in a published guide states which locales an instance
is being maintained in - so the answer is read off the objects themselves. The locales in use are
the union of the tags the selection's own translations carry, which needs no system-settings read
and is honest on an instance nobody has configured: an instance with one Lao translation is being
maintained in Lao, whatever a settings page says. Against that set, an object is short a NAME
translation, and short a FORM_NAME translation wherever DHIS2 gives it a form name to translate.

READ ONCE PER RESOURCE KIND. The translation read is one request per scope surface - ten of them -
rather than one per object, and the selection is applied to what comes back rather than written into
a filter that would put a national instance's organisation-unit UIDs into a query string.

REPORTING ONLY. Nothing here writes to DHIS2, and nothing here offers to. Acting on a finding -
changing the name, the code, or the translation in the instance - is the next slice, and the
roadmap is where it is stated.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

from dhis2w_fhir.i18n import FORM_NAME_PROPERTY, NAME_PROPERTY, normalize_locale
from dhis2w_fhir.service import resolve_validation_scope, validate_instance_codes
from dhis2w_fhir.validation.schemas import SCOPE_SURFACE_FIELDS
from pydantic import BaseModel, ConfigDict, Field

if TYPE_CHECKING:
    from dhis2w_client import Dhis2Client
    from dhis2w_fhir.config import GenerateConfig
    from dhis2w_fhir.validation.schemas import ValidationFinding, ValidationScope

#: What a compiled run answers, in the words a first-time reader can act on. A compiled guide is a
#: directory of resources with no instance behind it, so there is no metadata to grade - and saying
#: "not supported" would leave somebody looking for a setting that does not exist.
COMPILED_RUN_REASON = (
    "This server is reading a compiled implementation guide from disk, so there is no DHIS2 instance "
    "behind it to check. Start the server with --live to read the instance this guide was generated "
    "from."
)

#: The DHIS2 field each finding category is about, so a row says which spelling to go and change.
#: `template-hostile-name` is the one category that can be about either of two fields, and
#: `_FIELD_LABELS_BY_MESSAGE_PREFIX` is what tells them apart.
FIELD_BY_CATEGORY: dict[str, str] = {
    "template-hostile-name": "name",
    "template-hostile-code": "code",
    "invalid-code": "code",
    "missing-code": "code",
    "duplicate-code": "code",
    "code-stem-refusal": "code",
    "code-stem-fallback": "code",
    "unmapped-tracked-entity-type": "tracked entity type mapping",
}

#: The field labels `d2w fhir validate` opens a name finding's message with, longest first.
#:
#: A data element and a tracked entity attribute each carry two spellings DHIS2 puts on a page - the
#: name every vocabulary displays, and the form name a question is asked under - and both are graded
#: under one category, because they cost the same build. Which of the two a finding is about is
#: stated at the head of its own message, which is where this reads it: `test_metadata_health.py`
#: pins the pair against a real run, so a wording change in the validator is a failing test here
#: rather than a column that quietly says the wrong field.
_FIELD_LABELS_BY_MESSAGE_PREFIX = ("form name", "name")

#: What each severity costs the project, in the terms `d2w fhir validate` grades by.
_COST_BY_SEVERITY: dict[str, str] = {
    "error": "The implementation guide build stops on this.",
    "warning": "The build finishes, and what it publishes for this object is degraded.",
    "info": "The build finishes, and publishes this object as it stands.",
}

#: What an out-of-scope finding costs, which is nothing: this project publishes nothing from the object.
_INSTANCE_SCOPE_COST = "This object is outside what this project publishes, so no build of it reads the object."

#: The fields one translation read asks for. `formName` rides along on every surface: DHIS2 answers
#: it on the two collections that have one and omits it everywhere else, exactly as it does for the
#: instance-wide sweep `d2w fhir validate` reads through.
_TRANSLATION_FIELDS = "id,name,formName,translations[locale,property,value]"


class MetadataHealthFinding(BaseModel):
    """One thing `d2w fhir validate` found about a DHIS2 object, as a row states it.

    Every field but `field` and `cost` is the validator's own, carried across unchanged - the
    severity it graded, the scope it graded against, and the sentence it wrote. The two that are
    added here are derived from those: which DHIS2 field the finding is about, and what the grade
    costs this project.
    """

    model_config = ConfigDict(frozen=True)

    severity: Literal["error", "warning", "info"]
    scope: Literal["selection", "instance"]
    category: str
    """The validator's own name for the kind of defect - `invalid-code`, `template-hostile-name`."""

    resource_type: str
    """The DHIS2 metadata collection the object belongs to, in DHIS2's own spelling."""

    uid: str
    name: str
    code: str | None = None
    field: str | None = None
    """The DHIS2 field at fault, or None where the category is about none of an object's own fields."""

    message: str
    """The exact problem, in the validator's own words - the same sentence the report file carries."""

    cost: str
    """What this grade costs the project, said in one sentence rather than as a severity word."""


class MetadataHealthCounts(BaseModel):
    """How many findings there are of each severity, over the whole answer."""

    model_config = ConfigDict(frozen=True)

    errors: int = 0
    warnings: int = 0
    infos: int = 0


class LocaleCoverage(BaseModel):
    """How much of the selection one locale covers: names translated, and form names translated."""

    model_config = ConfigDict(frozen=True)

    locale: str
    """The BCP-47 tag, normalised from the Java locale DHIS2 stores - `pt_BR` reaches here as `pt-BR`."""

    name_count: int = 0
    """Selected objects carrying a NAME translation in this locale."""

    form_name_count: int = 0
    """Selected objects that have a DHIS2 form name and carry a FORM_NAME translation in this locale."""


class TranslationGap(BaseModel):
    """One selected object, and the locales in use it holds no translation for."""

    model_config = ConfigDict(frozen=True)

    resource_type: str
    uid: str
    name: str
    missing_name_locales: list[str] = Field(default_factory=list)
    missing_form_name_locales: list[str] = Field(default_factory=list)
    """Stated only for an object DHIS2 gives a form name - everything else has no form name to translate."""


class TranslationCoverage(BaseModel):
    """How far the selection is translated, per locale and per object.

    `locales` is the union of the tags the selection's own translations carry, which is what "in use
    on this instance" means here: an instance is being maintained in the languages somebody has
    written into it, and no system setting states that more honestly than the objects do. An empty
    list is an instance nobody has translated, in which case there are no gaps either - a gap is a
    locale another object already has and this one does not.
    """

    model_config = ConfigDict(frozen=True)

    locales: list[str] = Field(default_factory=list)
    object_count: int = 0
    """Selected objects the translation read covered."""

    form_named_count: int = 0
    """Of those, how many DHIS2 gives a form name - the denominator the form-name counts are read against."""

    per_locale: list[LocaleCoverage] = Field(default_factory=list)
    gaps: list[TranslationGap] = Field(default_factory=list)
    """One entry per selected object short of a translation, in resource type then name order."""


class MetadataHealth(BaseModel):
    """The whole answer: whether this run could look, what it found, and how far the selection is translated.

    `available` false is a compiled run and nothing else. It is answered as a body rather than as a
    refusal because it is a state a screen renders in words - there is nothing wrong with serving a
    compiled guide, and a page that read a 4xx here would have to invent the sentence this carries.
    """

    model_config = ConfigDict(frozen=True)

    available: bool = True
    reason: str | None = None
    """Why there is nothing to report, stated only when `available` is false."""

    graded_under: str | None = None
    """The `[generate] hostile_names` posture the severities were graded under, in the validator's own line."""

    object_count: int = 0
    """Metadata objects the validator swept, across every collection the instance holds."""

    counts: MetadataHealthCounts = Field(default_factory=MetadataHealthCounts)
    findings: list[MetadataHealthFinding] = Field(default_factory=list)
    translations: TranslationCoverage = Field(default_factory=TranslationCoverage)


class TranslatedObject(BaseModel):
    """One selected DHIS2 object as the translation read projects it - what it is called, and in what locales.

    The step between the wire and the coverage: `read_selected_translations` builds these off what
    DHIS2 answered, and `translation_coverage` is a pure function over them, so the arithmetic every
    number on the page is made of is testable without a request.
    """

    model_config = ConfigDict(frozen=True)

    resource_type: str
    uid: str
    name: str
    form_named: bool
    """Whether DHIS2 gives this object a form name - so whether it has one to be short a translation of."""

    name_locales: frozenset[str]
    """The normalised tags this object holds a NAME translation in."""

    form_name_locales: frozenset[str]
    """The normalised tags this object holds a FORM_NAME translation in, empty where it has no form name."""


def compiled_run_health() -> MetadataHealth:
    """What a run serving a compiled guide answers: nothing found, and the reason there is nothing."""
    return MetadataHealth(available=False, reason=COMPILED_RUN_REASON)


async def read_metadata_health(client: Dhis2Client, config: GenerateConfig) -> MetadataHealth:
    """Grade the instance behind this run: the validate findings over the selection, and its translations.

    The selection is resolved once and read by both halves - the validator grades severity against
    it, and the translation read narrows to it - so a national instance is scoped by one set of
    small reads rather than by two.
    """
    scope = await resolve_validation_scope(client, config)
    report = await validate_instance_codes(client, config, scope=scope)
    objects = await read_selected_translations(client, scope)
    return MetadataHealth(
        graded_under=report.hostile_names_line,
        object_count=report.object_count,
        counts=MetadataHealthCounts(errors=report.error_count, warnings=report.warning_count, infos=report.info_count),
        findings=[health_finding(finding) for finding in report.findings],
        translations=translation_coverage(objects),
    )


def health_finding(finding: ValidationFinding) -> MetadataHealthFinding:
    """One validate finding as a row states it: its own grade and wording, plus the field and the cost."""
    return MetadataHealthFinding(
        severity=finding.severity,
        scope=finding.scope,
        category=finding.category,
        resource_type=finding.resource_type,
        uid=finding.uid,
        name=finding.name,
        code=finding.code,
        field=field_at_fault(finding),
        message=finding.message,
        cost=cost_of(finding),
    )


def field_at_fault(finding: ValidationFinding) -> str | None:
    """Which DHIS2 field one finding is about, or None where the category is about no field of the object."""
    if finding.category == "template-hostile-name":
        return next(
            (label for label in _FIELD_LABELS_BY_MESSAGE_PREFIX if finding.message.startswith(f"{label} ")),
            "name",
        )
    return FIELD_BY_CATEGORY.get(finding.category)


def cost_of(finding: ValidationFinding) -> str:
    """What one finding's grade costs this project, in a sentence rather than in a severity word."""
    if finding.scope == "instance":
        return _INSTANCE_SCOPE_COST
    return _COST_BY_SEVERITY[finding.severity]


async def read_selected_translations(client: Dhis2Client, scope: ValidationScope) -> list[TranslatedObject]:
    """Read the translations the selection's objects carry, one bounded request per resource kind.

    Every collection a `ValidationScope` answers for is read whole and then narrowed to the scope,
    rather than filtered by a `id:in:[...]` a national organisation-unit selection would blow the
    query string with. A collection the selection holds nothing of costs no request at all.
    """
    objects: list[TranslatedObject] = []
    for resource_type in sorted(SCOPE_SURFACE_FIELDS):
        surface: frozenset[str] = getattr(scope, SCOPE_SURFACE_FIELDS[resource_type])
        if not surface:
            continue
        body = await client.get_raw(f"/api/{resource_type}", params={"fields": _TRANSLATION_FIELDS, "paging": "false"})
        objects.extend(_translated_objects(resource_type, body, surface))
    return objects


def translation_coverage(objects: list[TranslatedObject]) -> TranslationCoverage:
    """How far the selection is translated: the locales in use, the counts per locale, and every gap.

    "In use" is the union of the tags these objects carry, so an instance nobody has translated has
    no locales, no coverage rows, and no gaps - which is the honest reading of an instance in one
    language rather than a page full of everything being missing.
    """
    locales = sorted({locale for item in objects for locale in item.name_locales | item.form_name_locales})
    if not locales:
        return TranslationCoverage(object_count=len(objects), form_named_count=_form_named(objects))
    return TranslationCoverage(
        locales=locales,
        object_count=len(objects),
        form_named_count=_form_named(objects),
        per_locale=[
            LocaleCoverage(
                locale=locale,
                name_count=sum(1 for item in objects if locale in item.name_locales),
                form_name_count=sum(1 for item in objects if item.form_named and locale in item.form_name_locales),
            )
            for locale in locales
        ],
        gaps=sorted(
            (gap for gap in (_gap(item, locales) for item in objects) if gap is not None),
            key=lambda gap: (gap.resource_type, gap.name, gap.uid),
        ),
    )


def _form_named(objects: list[TranslatedObject]) -> int:
    """How many of the selected objects DHIS2 gives a form name to translate."""
    return sum(1 for item in objects if item.form_named)


def _gap(item: TranslatedObject, locales: list[str]) -> TranslationGap | None:
    """What one object is short of, or None when it is translated into every locale in use."""
    missing_names = [locale for locale in locales if locale not in item.name_locales]
    missing_form_names = (
        [locale for locale in locales if locale not in item.form_name_locales] if item.form_named else []
    )
    if not missing_names and not missing_form_names:
        return None
    return TranslationGap(
        resource_type=item.resource_type,
        uid=item.uid,
        name=item.name,
        missing_name_locales=missing_names,
        missing_form_name_locales=missing_form_names,
    )


def _translated_objects(resource_type: str, body: dict[str, Any], surface: frozenset[str]) -> list[TranslatedObject]:
    """Project one collection's wire answer onto the selected objects it holds, translations folded per property."""
    entries = body.get(resource_type)
    if not isinstance(entries, list):
        return []
    projected: list[TranslatedObject] = []
    for entry in entries:
        if not isinstance(entry, dict):
            continue
        uid = entry.get("id")
        if not isinstance(uid, str) or uid not in surface:
            continue
        form_name = entry.get("formName")
        name = entry.get("name")
        projected.append(
            TranslatedObject(
                resource_type=resource_type,
                uid=uid,
                name=name if isinstance(name, str) and name != "" else uid,
                form_named=isinstance(form_name, str) and form_name != "",
                name_locales=_locales_of(entry.get("translations"), NAME_PROPERTY),
                form_name_locales=_locales_of(entry.get("translations"), FORM_NAME_PROPERTY),
            )
        )
    return projected


def _locales_of(translations: object, property_name: str) -> frozenset[str]:
    """The normalised locale tags one object holds a translation of one property in."""
    if not isinstance(translations, list):
        return frozenset()
    return frozenset(
        normalize_locale(entry["locale"])
        for entry in translations
        if isinstance(entry, dict)
        and entry.get("property") == property_name
        and isinstance(entry.get("locale"), str)
        and entry.get("locale") != ""
        and isinstance(entry.get("value"), str)
        and entry.get("value") != ""
    )
